fastcrypto/src/vdf/wesolowski.rs

```diff
@@ -0,0 +1,216 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::error::FastCryptoError::{InvalidInput, InvalidProof};
+use crate::error::FastCryptoResult;
+use crate::groups::class_group::{Discriminant, QuadraticForm};
+use crate::groups::{ParameterizedGroupElement, UnknownOrderGroupElement};
+use crate::hash::{HashFunction, Sha256};
+use crate::vdf::VDF;
+use num_bigint::{BigInt, Sign};
+use num_integer::Integer;
+use num_prime::nt_funcs::is_prime;
+use std::cmp::min;
+use std::ops::Neg;
+
+/// Size of the random prime modulus B used in proving and verification.
+const B_BITS: usize = 264;
+
+/// An implementation of the Wesolowski VDF construction (https://eprint.iacr.org/2018/623) over a
+/// group of unknown order.
+pub struct WesolowskiVDF<G: ParameterizedGroupElement + UnknownOrderGroupElement> {
+    group_parameter: G::ParameterType,
+    iterations: u64,
+}
+
+impl<G: ParameterizedGroupElement + UnknownOrderGroupElement> WesolowskiVDF<G> {
+    /// Create a new VDF using the group defined by the given group parameter. Evaluating this VDF
+    /// will require computing `2^iterations * input` which requires `iterations` group operations.
+    fn new(group_parameter: G::ParameterType, iterations: u64) -> Self {
+        Self {
+            group_parameter,
+            iterations,
+        }
+    }
+}
+
+impl<G: ParameterizedGroupElement<ScalarType = BigInt> + UnknownOrderGroupElement> VDF
+    for WesolowskiVDF<G>
+{
+    type InputType = G;
+    type OutputType = G;
+    type ProofType = G;
+
+    fn evaluate(&self, input: &G) -> FastCryptoResult<(G, G)> {
+        if !input.has_group_parameter(&self.group_parameter) {
+            return Err(InvalidInput);
+        }
+
+        if self.iterations == 0 {
+            return Ok((input.clone(), G::zero(&self.group_parameter)));
+        }
+
+        let mut output = input.clone();
+        for _ in 0..self.iterations {
+            output = output.clone() + &output;
+        }
+
+        let challenge = get_challenge(&input.as_bytes(), &output.as_bytes());
+
+        // Algorithm from page 3 on https://crypto.stanford.edu/~dabo/pubs/papers/VDFsurvey.pdf
+        let two = BigInt::from(2);
+        let mut quotient_remainder = two.div_mod_floor(&challenge);
+        let mut proof = input.mul(&quotient_remainder.0);
+        for _ in 1..self.iterations {
+            quotient_remainder = (&quotient_remainder.1 * &two).div_mod_floor(&challenge);
+            proof = proof.double() + &input.mul(&quotient_remainder.0);
+        }
+
+        Ok((output, proof))
+    }
+
+    fn verify(&self, input: &G, output: &G, proof: &G) -> FastCryptoResult<()> {
+        if !input.has_group_parameter(&self.group_parameter)
+            || !output.has_group_parameter(&self.group_parameter)
+            || !proof.has_group_parameter(&self.group_parameter)
+        {
+            return Err(InvalidInput);
+        }
+
+        let challenge = get_challenge(&input.as_bytes(), &output.as_bytes());
```

Comment:
doesn't it include also the group parameters?

Reply:
No, but it could include a bit_length parameter which is right now given as a constant.